File: 2019W08/PreppindataW8.py
import pandas as pd

xls = pd.ExcelFile("E:\Week 8 Input.xlsx")
theft = pd.read_excel(xls,"Theft Audit") 
branch = pd.read_excel(xls,"Branch ID") 
theft["Type"] = theft["Type"].str.replace('^L.*','Liquid',regex=True)
theft["Type"] = theft["Type"].str.replace('.*Bar.*','Bar',regex=True)

temp = branch["Branch ID"].str.split(" - ", n = 1, expand = True) 
branch["Store ID"] = temp[0]
branch["Branch Name"] = temp[1]

final = theft.merge(branch, on="Store ID", how='inner')
final = final.pivot_table(index = ["Branch Name", "Crime Ref Number","Type"],
                          columns="Action",
                          values=["Date","Quantity"],
                          aggfunc={"Date":'max',"Quantity":'sum'})

final.columns=['_'.join(str(s).strip() for s in col if s) for col in final.columns]
final["Quantity_Stock Adjusted"].fillna(0,inplace = True)

final["Days to comp Adj"] = final.apply(lambda x: x["Date_Stock Adjusted"]-x["Date_Theft"],axis=1) 
final["Quantity Variance"] = final.apply(lambda x:  x["Quantity_Stock Adjusted"]+x["Quantity_Theft"],axis=1) 

final.rename(columns={'Date_Stock Adjusted':'Stock Adjusted',
                      'Date_Theft':'Theft',
                      'Quantity_Theft':'Stolen Volume'}, 
                 inplace=True)
final.drop("Quantity_Stock Adjusted",axis=1,inplace = True)
final.reset_index(inplace = True)

File: 2019W08/preppindataw8.R
library(readxl)
library(dplyr)
library(tidyr)

theft <- read_excel("E:/Week 8 Input.xlsx",
                    sheet = "Theft Audit", col_types = c("text", 
                                                         "text", "date", "numeric", "text", 
                                                         "text")) %>%
         mutate("Type" = if_else(grepl('.*Bar.*',`Type`), 'Bar', `Type`)) %>%
         mutate("Type" = if_else(grepl('^L.*',`Type`), 'Liquit', `Type`))  

branch <- read_excel("E:/Week 8 Input.xlsx", 
                     sheet = "Branch ID") %>%
          separate(`Branch ID`, c("Store ID", "Branch Name"), sep = " - ")

final <- theft %>%
      merge(., branch, by="Store ID") %>%
      select(-("Store ID")) %>%
      mutate_at(vars("Date","Quantity"), as.character) %>%
      gather(., key="var", value="value", "Date", "Quantity") %>%
      unite(newcol, `var`, `Action`) %>%
      spread(newcol, value) %>%
      mutate_at(vars(matches("Date")), as.Date) %>%
      mutate_at(vars(matches("Quantity")), as.numeric) %>%
      mutate_if(is.numeric, ~replace(., is.na(.), 0)) %>% 
      mutate("Days to cmop adj" = `Date_Stock Adjusted`-`Date_Theft`,
             "Stock Variance" =`Quantity_Stock Adjusted`+`Quantity_Theft`) %>%
      rename("Stock Adjusted" = "Date_Stock Adjusted",
             "Theft" = "Date_Theft",
             "Stolen Volume" = "Quantity_Theft") %>%
      select(-("Quantity_Stock Adjusted"))
  
View(final)

